app/reconciliation.go

```diff
@@ -0,0 +1,148 @@
+package app
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+)
+
+const reconStoreKey = "reconciliation"
+
+// unbalancedBlockHeightKey for saving unbalanced block height for reconciliation
+var unbalancedBlockHeightKey = []byte("0x01")
+
+var (
+	SupplyKey          = banktypes.SupplyKey
+	DenomAddressPrefix = banktypes.DenomAddressPrefix
+	BalancesPrefix     = banktypes.BalancesPrefix
+)
+
+// reconBalance will do reconciliation for accounts balances.
+func (app *App) reconBalance(ctx sdk.Context, bankIavl *iavl.Store) {
+	if ctx.BlockHeight() <= 2 {
+		return
+	}
+
+	height, exists := app.getUnbalancedBlockHeight(ctx)
+	if exists {
+		panic(fmt.Sprintf("unbalanced state at block height %d, please use hardfork to bypass it", height))
+	}
+
+	balanced := app.getBankChanges(ctx, bankIavl)
+	bankIavl.ResetDiff()
+
+	if !balanced {
+		app.saveUnbalancedBlockHeight(ctx)
+	}
+}
+
+func (app *App) getBankChanges(ctx sdk.Context, bankIavl *iavl.Store) bool {
+	supplyPre := sdk.Coins{}
+	balancePre := sdk.Coins{}
+	supplyCurrent := sdk.Coins{}
+	balanceCurrent := sdk.Coins{}
+
+	diff := bankIavl.GetDiff()
+	version := ctx.BlockHeight() - 2
```

Comment:
For a state sync node, it is possible to fail here as it does not have state of `ctx.BlockHeight() - 2`

Reply:
it seems there is a check above:

`	if ctx.BlockHeight() <= 2 {
		return
	}`

Reply:
For state sync, it will first fetch block N, then executes block N + 1. When running block N + 1, it actually recons block N, the diff is empty, so it will not trying to get store at version N - 1. Then it will not panic.

Reply:
Example
```
{"level":"info","module":"server","module":"statesync","height":870,"format":3,"chunk":0,"total":1,"time":"2023-07-05T20:05:13+08:00","message":"Applied snapshot chunk to ABCI app"}
{"level":"info","module":"server","module":"statesync","height":870,"appHash":"n-{\ufffdM|UB\ufffd1/\u0004\ufffd\ufffd\ufffd?\u001bp\u000b\ufffd\u001a6S3 \ufffdH}\ufffdgx\ufffd","time":"2023-07-05T20:05:13+08:00","message":"Verified ABCI app"}
{"level":"info","module":"server","module":"statesync","height":870,"format":3,"hash":"\ufffd;adN\u0010\ufffd\ufffd\ufffd\ufffd5T\u0016ɀ\ufffd\ufffd\ufffd\ufffd;\ufffdZ\ufffdc\ufffd\ufffd\"\ufffd \ufffd\ufffd\ufffd","time":"2023-07-05T20:05:13+08:00","message":"Snapshot restored"}
{"level":"info","module":"server","module":"blockchain","msg":{},"impl":"BlockPool","time":"2023-07-05T20:05:13+08:00","message":"service start"}
{"level":"info","module":"server","module":"blockchain","height":871,"time":"2023-07-05T20:05:14+08:00","message":"Time to switch to consensus reactor!"}
{"level":"info","module":"server","module":"blockchain","msg":{},"impl":{"Logger":{"Logger":{}}},"time":"2023-07-05T20:05:14+08:00","message":"service stop"}
{"level":"info","module":"server","module":"consensus","time":"2023-07-05T20:05:14+08:00","message":"SwitchToConsensus"}
{"level":"info","module":"server","module":"consensus","msg":{},"impl":"ConsensusState","time":"2023-07-05T20:05:14+08:00","message":"service start"}
{"level":"info","module":"server","module":"consensus","wal":"/Users/liguo/go/src/github.com/bnb-chain/greenfield/deployment/localup/.local/dataseed0/data/cs.wal/wal","msg":{},"impl":"baseWAL","time":"2023-07-05T20:05:14+08:00","message":"service start"}
{"level":"info","module":"server","module":"consensus","wal":"/Users/liguo/go/src/github.com/bnb-chain/greenfield/deployment/localup/.local/dataseed0/data/cs.wal/wal","msg":{},"impl":"Group","time":"2023-07-05T20:05:14+08:00","message":"service start"}
{"level":"info","module":"server","module":"consensus","msg":{},"impl":"TimeoutTicker","time":"2023-07-05T20:05:14+08:00","message":"service start"}
{"level":"info","module":"server","module":"consensus","height":871,"commit_round":0,"proposal":{},"commit":"E09EFCBA800B36DC626874E52264ED33A010FFD8551361E6E7F2DFA45E8BA197","time":"2023-07-05T20:05:14+08:00","message":"commit is for a block we do not know about; set ProposalBlock=nil"}
{"level":"info","module":"server","module":"consensus","height":871,"hash":"E09EFCBA800B36DC626874E52264ED33A010FFD8551361E6E7F2DFA45E8BA197","time":"2023-07-05T20:05:14+08:00","message":"received complete proposal block"}
{"level":"info","module":"server","module":"consensus","height":871,"hash":{},"root":"6E2D7B934D7C55429E312F04AFE5F13F1B700B961A36533320BB487DA067789E","num_txs":0,"time":"2023-07-05T20:05:14+08:00","message":"finalizing commit of block"}
{"level":"info","module":"server","supplyCurrent":[],"supplyPre":[],"balanceCurrent":[],"balancePre":[],"supplyChanges":[],"balanceChanges":[],"height":871,"version":869,"time":"2023-07-05T20:05:14+08:00","message":"reconciliation change details"}
```